.ci/magician/cmd/test_eap_vcr.go

```diff
@@ -0,0 +1,250 @@
+package cmd
+
+import (
+	_ "embed"
+	"fmt"
+	"magician/exec"
+	"magician/provider"
+	"magician/vcr"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+var tevEnvironmentVariables = [...]string{
+	"GEN_PATH",
+	"GOCACHE",
+	"GOPATH",
+	"GOOGLE_REGION",
+	"GOOGLE_ZONE",
+	"ORG_ID",
+	"GOOGLE_PROJECT",
+	"GOOGLE_BILLING_ACCOUNT",
+	"GOOGLE_ORG",
+	"GOOGLE_ORG_DOMAIN",
+	"GOOGLE_PROJECT_NUMBER",
+	"GOOGLE_USE_DEFAULT_CREDENTIALS",
+	"GOOGLE_IMPERSONATE_SERVICE_ACCOUNT",
+	"KOKORO_ARTIFACTS_DIR",
+	"HOME",
+	"MODIFIED_FILE_PATH",
+	"PATH",
+	"USER",
+}
+
+var testEAPVCRCmd = &cobra.Command{
+	Use:   "test-eap-vcr",
+	Short: "Run vcr tests for affected packages in EAP",
+	Long: `This command runs on new change lists to replay VCR cassettes and re-record failing cassettes.
+
+It expects the following arguments:
+	1. Change number
+
+
+The following environment variables are required:
+` + listTEVEnvironmentVariables(),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		env := make(map[string]string, len(tevEnvironmentVariables))
+		for _, ev := range tevEnvironmentVariables {
+			val, ok := os.LookupEnv(ev)
+			if !ok {
+				return fmt.Errorf("did not provide %s environment variable", ev)
+			}
+			env[ev] = val
+		}
+		rnr, err := exec.NewRunner()
+		if err != nil {
+			return err
+		}
+		vt, err := vcr.NewTester(env, "ci-vcr-cassettes", "ci-vcr-logs", rnr)
+		if err != nil {
+			return err
+		}
+		return execTestEAPVCR(args[0], env["GEN_PATH"], env["KOKORO_ARTIFACTS_DIR"], env["MODIFIED_FILE_PATH"], rnr, vt)
+	},
+}
+
+func listTEVEnvironmentVariables() string {
+	var result string
+	for i, ev := range tevEnvironmentVariables {
+		result += fmt.Sprintf("\t%2d. %s\n", i+1, ev)
+	}
+	return result
+}
+
+func execTestEAPVCR(changeNumber, genPath, kokoroArtifactsDir, modifiedFilePath string, rnr ExecRunner, vt *vcr.Tester) error {
+	vt.SetRepoPath(provider.Private, genPath)
+	if err := rnr.PushDir(genPath); err != nil {
+		return fmt.Errorf("error changing to gen path: %w", err)
+	}
+
+	changedFiles, err := rnr.Run("git", []string{"diff", "--name-only"}, nil)
+	if err != nil {
+		return fmt.Errorf("error diffing gen path: %w", err)
+	}
+
+	services, runFullVCR := modifiedPackages(strings.Split(changedFiles, "\n"), provider.Private)
+	if len(services) == 0 && !runFullVCR {
+		fmt.Println("Skipping tests: No go files or test fixtures changed")
+		return nil
+	}
+	fmt.Println("Running tests: Go files or test fixtures changed")
+
+	head := "auto-cl-" + changeNumber
+	if err := vt.FetchCassettes(provider.Private, "main", head); err != nil {
+		return fmt.Errorf("error fetching cassettes: %w", err)
+	}
+	replayingResult, testDirs, replayingErr := runReplaying(runFullVCR, provider.Private, services, vt)
+	if err := vt.UploadLogs(vcr.UploadLogsOptions{
```

Comment:
Since the EAP jobs run within an environment where we have logging accessible the way we want, do we still need to upload logs the same way? I could definitely see individual debug logs per test being useful, but was not sure about other logging.

Reply:
I'd prefer to have the same upload logs function for both environments in addition to any logging done by Kokoro.